streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSource.java

```diff
@@ -101,8 +99,17 @@ public void process(final K key, final V value) {
             }
 
             if (queryableName != null) {
-                final ValueAndTimestamp<V> oldValueAndTimestamp = sendOldValues ? store.get(key) : null;
-                final V oldValue = getValueOrNull(oldValueAndTimestamp);
+                final ValueAndTimestamp<V> oldValueAndTimestamp = store.get(key);
+                final V oldValue;
+                if (oldValueAndTimestamp != null) {
+                    oldValue = oldValueAndTimestamp.value();
+                    if (context().timestamp() < oldValueAndTimestamp.timestamp()) {
+                        LOG.warn("Detected out-of-order KTable update for {} at offset {}, partition {}.",
```

Comment:
I know I suggested this, but just clarifying: in the task's record queue, we already detect lateness like this and also record in the `record-lateness` metric already, so here we just log it for KTable source specifically in addition to the general metrics.

Reply:
Well. We record `record-lateness` in `PartitionGroup`. Here we don't record any metric but only issue a WARN log. In general, out-of-order data are ok, but for this case it's a correctness concern, and hence the warning.

I would actually like to take this one step further (in a follow up KIP), to allow people to opt-in and drop those records.